File: Data Structures & Algorithms/search-2d-matrix/submission-11.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        ROWS = len(matrix)
        COLS = len(matrix[0])
        left, right = 0, ROWS * COLS - 1

        while left <= right:
            mid = left + (right - left) // 2
            rm = mid // COLS
            cm = mid % COLS

            if matrix[rm][cm] == target:
                return True

            elif matrix[rm][cm] < target:
                left = mid + 1
            else:
                right = mid - 1
        

        return False

File: Data Structures & Algorithms/eating-bananas/submission-11.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int left = 1;
        int right = 0;

        for (int pile : piles) {
            right = Math.max(pile, right);
        }

        int res = right;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            int hour = 0;
            for (int pile : piles) {
                hour += (int) Math.ceil((double) pile / mid);
            }

            if (hour <= h) {
                res = Math.min(res, mid);
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return res;
    }
}
